File: backend/internal/modules/category/category_service.go
package category

import (
	"citramascoweb-backend/internal/dto"
	"strings"
	"time"

	"github.com/google/uuid"
)

type categoryService struct {
	categoryRepo CategoryRepositoryInterface
}

func Slugify(text string) string {
	text = strings.ToLower(text)
	text = strings.ReplaceAll(text, " ", "-")
	return text
}

func NewCategoryService(categoryRepo CategoryRepositoryInterface) *categoryService {
	return &categoryService{categoryRepo: categoryRepo}
}

func (s *categoryService) FindById(id string) (*Category, error) {
	category, err := s.categoryRepo.GetById(id)

	if err != nil {
		return nil, err
	}

	return category, nil
}

func (s *categoryService) FindAll() ([]Category, error) {
	categories, err := s.categoryRepo.GetAll()

	if err != nil {
		return nil, err
	}

	return categories, nil
}

func (s *categoryService) FindBySlug(slug string) (*Category, error) {
	category, err := s.categoryRepo.GetBySlug(slug)

	if err != nil {
		return nil, err
	}

	return category, nil
}

func (s *categoryService) Create(req *dto.CreateCategoryRequest) error {

	now := time.Now()
	category := &Category{
		Id:        uuid.New().String(),
		Name:      req.Name,
		Slug:      Slugify(req.Name),
		CreatedAt: now,
		IsDeleted: false,
	}

	err := s.categoryRepo.Store(category)
	if err != nil {
		return err
	}

	return nil
}

func (s *categoryService) Update(id string, req *dto.UpdateCategoryRequest) error {

	category, err := s.categoryRepo.GetById(id)
	if err != nil {
		return err
	}

	category.Name = req.Name
	category.Slug = Slugify(req.Name)

	err = s.categoryRepo.Update(id, category)
	if err != nil {
		return err
	}

	return nil

}

func (s *categoryService) Delete(id string) error {
	err := s.categoryRepo.Delete(id)

	if err != nil {
		return err
	}

	return nil
}

File: frontend/src/services/categoryService.ts
import api from './api'
import type { ApiResponse, Category } from '@/types'

export const categoryService = {
  /**
   * GET /api/categories — Fetch all categories
   */
  getAll: async (): Promise<Category[]> => {
    const response = await api.get<ApiResponse<Category[]>>('/categories')
    return response.data.data ?? []
  },

  /**
   * GET /api/category/:slug — Fetch category by slug
   */
  getBySlug: async (slug: string): Promise<Category> => {
    const response = await api.get<ApiResponse<Category>>(`/category/${slug}`)
    if (!response.data.data) {
      throw new Error(response.data.message || 'Category not found')
    }
    return response.data.data
  },

  /**
   * GET /api/category/detail/:id — Fetch category by ID
   */
  getById: async (id: string): Promise<Category> => {
    const response = await api.get<ApiResponse<Category>>(`/category/detail/${id}`)
    if (!response.data.data) {
      throw new Error(response.data.message || 'Category not found')
    }
    return response.data.data
  },

  /**
   * POST /api/category/ — Create a new category
   */
  create: async (data: { name: string }): Promise<string> => {
    const response = await api.post<ApiResponse<any>>('/category/', data)
    if (!response.data.success) {
      throw new Error(response.data.message || 'Validation failed!')
    }
    return response.data.message
  },

  /**
   * PATCH /api/category/:id — Update a category
   */
  update: async (id: string, data: { name: string }): Promise<string> => {
    const response = await api.patch<ApiResponse<any>>(`/category/${id}`, data)
    if (!response.data.success) {
      throw new Error(response.data.message || 'Validation failed!')
    }
    return response.data.message
  },

  /**
   * DELETE /api/category/:id — Delete a category
   */
  delete: async (id: string): Promise<string> => {
    const response = await api.delete<ApiResponse<any>>(`/category/${id}`)
    if (!response.data.success) {
      throw new Error(response.data.message || 'Deletion failed!')
    }
    return response.data.message
  },
}
